include/nbl/ext/TextRendering/TextRendering.h

```diff
@@ -0,0 +1,126 @@
+
+	TextRenderer(core::smart_refctd_ptr<ILogicalDevice> device, uint32_t maxGlyphCount, uint32_t maxStringCount, uint32_t maxGlyphsPerString);
+
+	using glyph_geometry_pool_t = video::SubAllocatedDataBufferST<>;
+	// Data stored as SoA in property pool:
+	// - Glyph offset
+	// - String bounding box
+	using string_pool_t = video::CPropertyPool<core::allocator, glyph_geometry_pool_t::size_type, StringBoundingBox>;
+
+	struct string_handle_t
+	{
+		string_pool_t::size_type stringAddr = string_pool_t::invalid_address;
+		glyph_geometry_pool_t::size_type glyphDataAddr = glyph_geometry_pool_t::invalid_address;
+		glyph_geometry_pool_t::size_type glyphCount = 0u;
+	};
+
+	template<class Clock = typename std::chrono::steady_clock>
+	uint32_t allocateStrings(
+		const std::chrono::time_point<Clock>& maxWaitPoint, // lets have also a version without this (default)
+		const uint32_t count, // how many strings
+		string_handle_t* handles, // output handles, if `glyphDataAddr` was not primed with invalid_address, allocation will not happen, likewise for `stringDataAddr`
+		const char* const* stringData,
+		const StringBoundingBox* wrappingBoxes = nullptr // optional, to wrap paragraphs
+	);
+
+	inline uint32_t allocateStrings(
+		const uint32_t count, // how many strings
+		string_handle_t* handles, // output handles, if `glyphDataAddr` was not primed with invalid_address, allocation will not happen, likewise for `stringDataAddr`
+		const char* const* stringData,
+		const StringBoundingBox* wrappingBoxes = nullptr // optional, to wrap paragraphs
+	)
+	{
+		return allocateStrings(GPUEventWrapper::default_wait(), count, handles, stringData, wrappingBoxes);
+	}
+
+	void freeStrings(
+		const uint32_t count, // how many strings
+		const string_handle_t* handles,
+		core::smart_refctd_ptr<IGPUFence>&& fence // for a deferred free
+	);
+
+	// Creates the pipeline using the cached pipeline layout
+	core::smart_refctd_ptr<video::IGPUGraphicsPipeline> createPipeline(video::IGPUObjectFromAssetConverter::SParams& cpu2gpuParams);
+
+	// The data in these buffers is provided by the frustum culling system before calling drawText
+	void updateDescriptorSet(
+		core::smart_refctd_ptr<video::IGPUBuffer> visibleStrings,
+		core::smart_refctd_ptr<video::IGPUBuffer> visibleStringMvps,
+		core::smart_refctd_ptr<video::IGPUBuffer> visibleStringGlyphCount,
+		core::smart_refctd_ptr<video::IGPUBuffer> prefixSumOutput,
+		core::smart_refctd_ptr<video::IGPUBuffer> indirectDrawArgs
+	);
+
+	void drawText(core::smart_refctd_ptr<video::IGPUCommandBuffer> cmdbuf, core::smart_refctd_ptr<video::IGPUBuffer> indirectDrawArgs);
+
+	void drawTextIndexed(core::smart_refctd_ptr<video::IGPUCommandBuffer> cmdbuf, core::smart_refctd_ptr<video::IGPUBuffer> indirectDrawArgs);
+
+private:
+
+private:
+	core::smart_refctd_ptr<ILogicalDevice> m_device;
+
+	core::smart_refctd_ptr<video::IGPUPipelineLayout> m_pipelineLayout;
+	core::smart_refctd_ptr<video::IGPUDescriptorSet> m_descriptorSet;
```

Comment:
only cache (keep global) the default layouts, indexed/non-indexed pipeline and descriptor set with the two pools.

Let user provide the buffer & descriptor set for the prefix summing / visible string list.

Reply:
actually dont cache the pipeline layout or pipeline (make create functions), user will need to override the shaders